File: JS/栈与队列/20_有效的括号.js
var isValid = function(s) {
    let map = new Map([
        [')', '('],
        ['}', '{'],
        [']', '[']
    ])
    let stack = []
    for(let str of s){
        if(map.has(str)){
            if(!stack.length || stack[stack.length-1] !== map.get(str)){
                return false
            }
            stack.pop()
        }else{
            stack.push(str)
        }
    }
    return !stack.length
};

s = "()"
console.log(isValid(s))

File: Python/面试经典150题/栈/20_有效的括号.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        map = {
            ')': '(',
            '}': '{',
            ']': '['
        }
        for str in s:
            if len(stack) == 0:
                stack.append(str)
                continue
            if str in [']','}',')'] and map[str] == stack[-1]:
                stack.pop()
            else:
                stack.append(str)
        return len(stack) == 0

s = "([])"
print(Solution().isValid(s))